### PR TITLE
refactor: Integrate AVS API

### PR DESCRIPTION
**Pull request by atomrc**
TODOs:
- [x] restore screensharing for the wrapper (with electron's desktopCapturer)
- [x] restore routing the audio from a call to the selected output device
- [x] handle upgrading from audio to video
- [x] Handle media stream change on the UI side (when the user chooses another input device while in a call)
- [x] Show video preview when an incoming video call is ringing 
- [x] Show video preview when outgoing video **group** call is started 
- [x] Check that group > 4 ppl cannot do video call
- [x] Listen to mute state changes in avs
- [x] Save call start/call end events in the database
- [x] Find a way to get the media streams from the avs lib (in order to display the video element on screen)
- [x] adapt the video grid handler with the new data format
- [x] Get the connected participants from avs lib
- [x] ~~Ability to remove a participant while in a call (should also be removed from avs)~~
- [x] Close ongoing call when the app is reloaded
- [x] Handle group calls

TODO (before merge):
- [x] restore the unit test in `package.json`;
- [x] remove deployment script to `-qa` environment.
- [x] Restore the build script to mode prod

### CODE REVIEW

File: src/script/calling/videoGridHandler.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Will compute the next grid layout according to the previous state and the new array of streams
+ * The grid will fill according to this pattern
+ * - 1 stream : [id, '', '', '']
+ * - 2 streams: [id, '', id, '']
+ * - 3 streams: [id, '', id, id]
+ * - 3 streams: [id, id, '', id]
+ * - 4 streams: [id, id, id, id]
+ * @param {Array<string|0>} previousGrid - the previous state of the grid
```
Comment (ffflorian) [inline src/script/calling/videoGridHandler.ts]: JSDoc types are not needed in TypeScript.
```suggestion
 * @param previousGrid the previous state of the grid
```

File: src/script/calling/CallingRepository.ts [hunk 2]
```diff
@@ -0,0 +1,971 @@
...
+  STATE as CALL_STATE,
+  VIDEO_STATE,
+  Wcall,
+  getAvsInstance,
+} from '@wireapp/avs';
+
+export class CallingRepository {
+  private readonly backendClient: any;
+  private readonly conversationRepository: any;
+  private readonly eventRepository: any;
```
Comment (ffflorian) [inline src/script/calling/CallingRepository.ts:68]: Here you could just import `EventRepository` even though it's not converted to TypeScript yet.

File: src/script/calling/CallingRepository.ts [hunk 3]
```diff
@@ -0,0 +1,971 @@
...
+  private readonly serverTimeHandler: any;
+
+  private selfUser: any;
+  private selfClientId: DeviceId;
+  private isReady: boolean = false;
+  private wUser: number | undefined;
+  private wCall: Wcall | undefined;
+  public readonly activeCalls: ko.ObservableArray<Call>;
+  private readonly isMuted: ko.Observable<boolean>;
+  private incomingCallCallback: Function;
```
Comment (ffflorian) [inline src/script/calling/CallingRepository.ts]: Can we specify this function a bit more? Something like
```ts
private incomingCallCallback: (data: ...) => void;
```

File: src/script/calling/CallingRepository.ts [hunk 4]
```diff
@@ -0,0 +1,971 @@
...
+
+  // will cache the query to media stream (in order to avoid asking the system for streams multiple times when we have multiple peers)
+  private mediaStreamQuery: Promise<MediaStream> | undefined;
+
+  public readonly joinedCall: ko.PureComputed<Call | undefined>;
+
+  private callingConfig: any;
+  private callingConfigTimeout: number | undefined;
```
Comment (ffflorian) [inline src/script/calling/CallingRepository.ts]:
```suggestion
  private callingConfigTimeout?: number;
```

Same goes for all other parameters which can be `undefined`.

File: src/script/calling/CallingRepository.ts [hunk 5]
```diff
@@ -0,0 +1,971 @@
...
+  public readonly activeCalls: ko.ObservableArray<Call>;
+  private readonly isMuted: ko.Observable<boolean>;
+  private incomingCallCallback: Function;
+
+  // will cache the query to media stream (in order to avoid asking the system for streams multiple times when we have multiple peers)
+  private mediaStreamQuery: Promise<MediaStream> | undefined;
+
+  public readonly joinedCall: ko.PureComputed<Call | undefined>;
+
+  private callingConfig: any;
```
Comment (ffflorian) [inline src/script/calling/CallingRepository.ts]: Can we type this?

File: test/unit_tests/calling/CallingRepositorySpec.js [hunk 6]
```diff
@@ -17,60 +17,384 @@
...
+      content: JSON.parse(payload),
+      conversation: conversationId,
+      from: userId,
+      sender: clientId,
+      time: Date.now(),
+    };
+    remoteCallingRepository.onCallEvent(event, EventRepository.SOURCE.WEB_SOCKET);
+  };
+
+  const incoming = conversationId => wCall.answer(wUser, conversationId, CALL_TYPE.AUDIO, 0);
```
Comment (bennycode) [inline test/unit_tests/calling/CallingRepositorySpec.js:379]:
```suggestion
  const incoming = conversationId => wCall.answer(wUser, conversationId, CALL_TYPE.NORMAL, 0);
```

File: src/script/view_model/CallingViewModel.ts [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+declare global {
+  interface HTMLAudioElement {
+    setSinkId: (sinkId: string) => Promise<void>;
+  }
+}
+
+export class CallingViewModel {
+  public readonly audioRepository: any;
+  public readonly callingRepository: CallingRepository;
+  public readonly conversationRepository: any;
```
Comment (ffflorian) [inline src/script/view_model/CallingViewModel.ts]: You could  import the repository and specify it here, even when the repository is not converted to TypeScript yet.

File: src/script/view_model/CallingViewModel.ts [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+      callingRepository.activeCalls().filter(call => call.reason() !== CALL_REASON.ANSWERED_ELSEWHERE),
+    );
+    this.selectableScreens = ko.observable([]);
+    this.isChoosingScreen = ko.pureComputed(() => this.selectableScreens().length > 0);
+    this.multitasking = multitasking;
+
+    this.onChooseScreen = () => {};
+
+    const ring = (call: Call): void => {
+      const sounds: any = {
```
Comment (ffflorian) [inline src/script/view_model/CallingViewModel.ts]: Why `any`?

**Pull request merged**